### ISSUE TITLE
Lerna and package-lock v2 problems

### ISSUE THREAD

**Issue by maribethb** [RaspberryPiFoundation/blockly-samples#862]
**Category**

Plugins

**Component**

Publishing

**Describe the bug**

Lerna and package-locks don't mix. We've known this for a while. Notably, you have to run `npm run clean:node && npm install` and then check for changes to package-locks before you can publish, because if there would be changes to package-locks during the publish step, the entire step fails because there will be uncommitted changes in your local git repository. 

However, it's gotten worse after updating to package-lock v2. This is because v2 has syntax that puts the package version number in two different places in the package-lock, and only one of them gets updated during the publish step. So after the publish step, the package-lock is out of sync with the version number of its own package. For example,

```json
"name": "@blockly/dev-tools",
	"version": "2.5.4",
	"lockfileVersion": 2,
	"requires": true,
	"packages": {
		"": {
			"name": "@blockly/dev-tools",
			"version": "2.5.3",
```

See https://github.com/google/blockly-samples/pull/859 for more. And #861 which is only a few hours later.

So the next time a package is published, there are guaranteed to be changes to the package-lock because `npm install` corrects this inconsistency.

And to make it worse, I think lerna takes this change to package-lock as a sign that the package needs to be republished. If you run `lerna changed` after running `npm run clean:node && npm install` then lerna reports every single package needs to be republished, because every single package just had its package-lock changed. And if you do this publish, you are now trapped in a loop where now the package-lock is inconsistent again and you need to update them...

I don't know how to fix this but I feel that this increases the urgency for fixing our lerna / package-lock problems because we will now be publishing packages unnecessarily every week.

**Additional context**

**Comment by maribethb** [comment]
First, fixing these issues with package-lock still needs to be done at some point in order to simplify the release process. Or if we move to release-please for releases then this won't be an issue, hopefully. 

But in the short term, I think we can fix this particular issue with two steps:

1. Add package-lock.json to the ignored paths for lerna. Lerna won't decide we need to republish if the only changes are to package-lock. See `command.publish.ignoreChanges` https://github.com/lerna/lerna#lernajson
2. After a release, run npm install to update the package-locks with the correct version number for any packages that were published. Commit any changes to package-lock.

1 should be safe because we don't actually include the package-locks with the published package, so there's no need to republish if that's the only change, iiuc.
2 is just to keep things tidy and the package-locks in a consistent state. Doing it after the release also might mean there won't be changes to it before the next release so it moves the chore to after the release instead of before, which might just be less stressful (waiting on review before merging, etc)

Note that lerna has an open issue about this: https://github.com/lerna/lerna/issues/2832 and they've had a PR open for it since May so I don't know if they plan to merge that soon. If they do, it should remove the need to do step 2, but I think it's fine to leave step 1 in place.

Also, thanks to Abby for helping me look into this and finding the linked lerna issue.

**Comment by samelhusseini** [comment]
Re: 1, does that mean lerna won't try to update the package-lock if it's added to the `ignoreChanges` as well? If not, won't you end up in a situation where everytime `npm run boot` is run it will have uncommitted changes to package-lock of lerna changes?

**Comment by maribethb** [comment]
I can test it, but I don't think so. I think that particular setting is only used to tell lerna "changes in this file don't require a new release" not that lerna should not modify it.

**Comment by samelhusseini** [comment]
Ok, so going ahead with 1 sounds good, but if you are to do 2, wouldn't have be more cumbersome to constantly have modified package-locks that you don't want to commit because they conflict with the ones generated by an `npm install` after a release?

**Comment by maribethb** [comment]
Hm, I'm not sure what you mean. My understanding is
1. During a release, lerna updates the package-lock incorrectly/incompletely as described in the issue
2. After the release, we immediately run npm install to fix the inconsistency in package lock and commit the result
3. During the normal course of development if there are other changes to package-lock, those would be committed as well like normal

So there wouldn't be any conflicts and you wouldn't ever have to not commit package-lock changes. Does that make sense or would there be another conflict I'm not seeing?

If we didn't do 2, then our package-locks would be in an inconsistent/nonsense state and they would just need to be updated whenever npm install happens to be run, instead of doing it routinely after the release. So if there is a problem here I think we'd run into it no matter what, it's just a question of when we want to resolve that inconsistency.

**Comment by samelhusseini** [comment]
Yeah (if my understanding is correct), I believe lerna **is** the one creating these inconsistent package-locks, so if you run `npm install` in each of the plugin directories after a release, then you get a valid package-lock.json and you can commit that. But during the normal course of development, when lerna runs `lerna bootstrap` it will create package-lock.jsons that are inconsistent with the one that `npm install` creates, thus leaving you in a state with modified package-locks but with modifications you don't want to commit.

**Comment by maribethb** [comment]
As far as I know, the package-locks are only inconsistent after a release. Because lerna updates the first instance of the version in package-lock when it updates a version, but not the second one (which is only a feature of package-lock v2 so it's just incompatible with npm v7). e.g.

```
"name": "@blockly/dev-tools",
	"version": "2.5.4", // this is the new version that lerna just published/updated
	"lockfileVersion": 2,
	"requires": true,
	"packages": {
		"": {
			"name": "@blockly/dev-tools",
			"version": "2.5.3", // lerna forgot to update this
```

so running `npm install` after an update, all it should do is correct this inconsistency so that we get

```
"name": "@blockly/dev-tools",
	"version": "2.5.4", // this is the new version that lerna just published/updated
	"lockfileVersion": 2,
	"requires": true,
	"packages": {
		"": {
			"name": "@blockly/dev-tools",
			"version": "2.5.4", // matches the above version
```

So I'm not seeing how doing this after the release would conflict with later changes. If we don't do it after the release, it will need to be done sometime: any time you run `npm install` normally, the package-lock will correct itself and this change will be made.

If you're saying that during normal development, the changes made to package-lock are actually invalid, then... I don't know about that, that seems like a separate issue we'd already be seeing? Or I'm still misunderstanding you.

**Comment by maribethb** [comment]
Sam and I had a chat and I learned about Lerna! I'm going to write a separate issue or doc about the full decision about package-locks, but for the context of this specific issue:

The miscommunication we were having is Sam meant that going into each plugin's directory and running `npm install` gives you a different package-lock result than running `npm install` at the root level of the repository (because at the root level [or more accurately, any level above an individual plugin], `npm install` runs `lerna boot` and it's the lerna bootstrap that gives the conflicting package.json). I was assuming we would just run `npm install` at the top level, which would give you the same result that lerna gives so there is no conflict.

SO the final outcome is that we can
1. Have lerna ignore `package-lock.json` when deciding if we need a new version
2. After release, run `npm install` in the root level of blockly-samples and commit the result (we wouldn't need to do this if we either change how we use lerna or if lerna fixes their underlying issue with not updating the version in both places)

Running `npm install` at the root level is not the same result as running it in the individual plugin, and that tension is the same issue we've already had. Sam helped me understand more about where that problem is coming from and I'll put more details into a separate issue/doc for discussion.
